### PR TITLE
solution checksum and serialization service.

### PR DESCRIPTION
**Pull request by heejaechang**
actual commit with changes.
https://github.com/dotnet/roslyn/pull/12771/commits/d0b58a20bc1586ff68bb49a1c1f23e56169662b9

work left for this area.
https://github.com/dotnet/roslyn/issues/12770

### CODE REVIEW

**Comment by heejaechang** [comment]
@mattwar @CyrusNajmabadi @dotnet/roslyn-analysis @dotnet/roslyn-ide @mavasani @srivatsn can you take a look? thank you.

**Comment by heejaechang** [comment]
@panopticoncentral adding paul.

**Comment by heejaechang** [comment]
test will fail since PR doesnt contain csproj/vbproj file changes for added files

File: src/Workspaces/Core/Portable/Execution/Serialization/SnapshotStorage.cs [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    /// <summary>
+    /// This is collection of hierarchical checksum trees
+    /// </summary>
+    internal class SnapshotStorages
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/SnapshotStorage.cs]: >SnapshotStorages [](start = 19, length = 16)

Should be SnapshotStorage?

File: src/Workspaces/Core/Portable/Execution/Extensions.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    // REVIEW: should there be auto retry mechanism such as if out of proc run fails, it automatically calls inProc one?
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Extensions.cs]: >// REVIEW: should there be auto retry mechanism such as if out of proc run fails, it automatically calls inProc one? [](start = 4, length = 116)

This comment doesn't seem to be connected to anything in this file?

File: src/Workspaces/Core/Portable/Execution/Serialization/IReferenceSerializationService.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+        MetadataReference ReadMetadataReferenceFrom(ObjectReader reader, CancellationToken cancellationToken);
+        AnalyzerReference ReadAnalyzerReferenceFrom(ObjectReader reader, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// This lets consumer to get to inner temporary storage that references use
+    /// as its shadow copy storage
+    /// </summary>
+    internal interface ISupportTemporaryStorage
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/IReferenceSerializationService.cs]: >ISupportTemporaryStorage [](start = 23, length = 24)

Shouldn't this be in its own file?

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+                    WriteMvidTo(module, writer, cancellationToken);
+                }
+
+                return;
+            }
+
+            WriteMvidTo((ModuleMetadata)metadata, writer, cancellationToken);
+        }
+
+        private void WriteMvidTo(ModuleMetadata metadata, ObjectWriter writer, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 79, length = 35)

Not used.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+            using (var writer = new ObjectWriter(stream, cancellationToken: cancellationToken))
+            {
+                WriteMvidsTo(reference, writer, cancellationToken);
+
+                stream.Position = 0;
+                return Checksum.Create(stream);
+            }
+        }
+
+        private void WriteMvidsTo(PortableExecutableReference reference, ObjectWriter writer, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 94, length = 35)

Not used once you remove the param from WriteMvidTo

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+
+                // TODO: deal with xml document provider properly
+                //       should be shadow copy xml doc comment?
+                return new SerializedMetadataReference(properties, filePath, tuple.Item1, tuple.Item2, XmlDocumentationProvider.Default);
+            }
+
+            throw ExceptionUtilities.UnexpectedValue(kind);
+        }
+
+        private void WriteTo(MetadataReferenceProperties properties, ObjectWriter writer, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 90, length = 35)

Not used.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+        {
+            writer.WriteInt32((int)properties.Kind);
+            writer.WriteArray(properties.Aliases.ToArray());
+            writer.WriteBoolean(properties.EmbedInteropTypes);
+        }
+
+        private MetadataReferenceProperties ReadMetadataReferencePropertiesFrom(ObjectReader reader, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 101, length = 35)

Not used.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+        private void CopyByteArrayToStream(ObjectReader reader, Stream stream, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // TODO: make reader be able to read byte[] chunk
+            var content = reader.ReadArray<byte>();
+            stream.Write(content, 0, content.Length);
+        }
+
+        private void WriteTo(ModuleMetadata metadata, ObjectWriter writer, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 75, length = 35)

Not used

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+        {
+            writer.WriteInt32((int)metadata.Kind);
+
+            var metadataReader = GetMetadataReader(metadata);
+            WriteTo(metadataReader, writer, cancellationToken);
+        }
+
+        private void WriteTo(MetadataReader reader, ObjectWriter writer, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 73, length = 35)

Not used.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+        private MetadataReferenceProperties ReadMetadataReferencePropertiesFrom(ObjectReader reader, CancellationToken cancellationToken)
+        {
+            var kind = (MetadataImageKind)reader.ReadInt32();
+            var aliases = reader.ReadArray<string>().ToImmutableArrayOrEmpty();
+            var embedInteropTypes = reader.ReadBoolean();
+
+            return new MetadataReferenceProperties(kind, aliases, embedInteropTypes);
+        }
+
+        private void WriteTo(Metadata metadata, ObjectWriter writer, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 69, length = 35)

Not used.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+
+        private void WritePortableExecutableReferenceTo(PortableExecutableReference reference, ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            WritePortableExecutableReferenceHeaderTo(reference, SerializationKinds.Bits, writer, cancellationToken);
+            WritePortableExecutableReferenceMetadataTo(reference, writer, cancellationToken);
+
+            // TODO: what I should do with documentation provider? it is not exposed outside
+        }
+
+        private void WritePortableExecutableReferenceMetadataTo(PortableExecutableReference reference, ObjectWriter writer, CancellationToken cancellationToken)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >CancellationToken cancellationToken [](start = 124, length = 35)

Not used.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+        {
+            // TODO: right now, use reflection, but this API will be added as public API soon. when that happen, remove reflection
+            var methodInfo = reference.GetType().GetTypeInfo().GetDeclaredMethod("GetMetadataImpl");
+            return (Metadata)methodInfo.Invoke(reference, null);
+        }
+
+        private sealed class PinnedObject : IDisposable
+        {
+            private readonly GCHandle _gcHandle;
+            private readonly long _length;
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: > _length [](start = 33, length = 8)

This field is never used.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+                _gcHandle = GCHandle.Alloc(array, GCHandleType.Pinned);
+                _length = length;
+            }
+
+            internal IntPtr GetPointer()
+            {
+                return _gcHandle.AddrOfPinnedObject();
+            }
+
+            private void Dispose(bool disposing)
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: >disposing [](start = 38, length = 9)

Parameter isn't used.

File: src/Workspaces/Core/Portable/Execution/Serialization/Asset.cs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        public override Task WriteToAsync(ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            _writer(_language, _value, writer, cancellationToken);
+            return SpecializedTasks.EmptyTask;
+        }
+    }
+
+    internal sealed class MetadataReferenceAsset : Asset
```
Comment (panopticoncentral) [inline src/Workspaces/Core/Portable/Execution/Serialization/Asset.cs]: >MetadataReferenceAsset [](start = 26, length = 22)

Aren't classes like these split out into their own files?

**Comment by panopticoncentral** [comment]
:+1: But there are specific questions in there that someone with more Roslyn experience than I have should answer or at least review...

File: src/Workspaces/Core/Portable/Execution/Serialization/SnapshotStorage.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/SnapshotStorage.cs]: renamed to ChecksumTreeCollections

File: src/Workspaces/Core/Portable/Execution/Extensions.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Extensions.cs]: yep. deleted

File: src/Workspaces/Core/Portable/Execution/Serialization/IReferenceSerializationService.cs (hunk 3 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/IReferenceSerializationService.cs]: done

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 5 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: made inner one to check cancellation

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 4 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: added cancellation check

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 11 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: now WriteTo check cancellation

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 6 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: added cancellation check

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 7 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: added cancellation check.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 10 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: same

File: src/Workspaces/Core/Portable/Execution/Serialization/Asset.cs (hunk 14 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/Asset.cs]: I have work item to clean up some assets. I will do that when I do that work.

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 8 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: same here, cancellation check is added to callee

File: src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs (hunk 9 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/Serialization/AbstractReferenceSerializationService.cs]: same

**Comment by heejaechang** [comment]
@panopticoncentral I addressed your feedback in this commit https://github.com/dotnet/roslyn/pull/12771/commits/2bc40863d782a9b44ef2e473ae29430a539272b6

**Comment by panopticoncentral** [comment]
:+1:

**Comment by heejaechang** [comment]
this PR is merged to this
https://github.com/dotnet/roslyn/pull/12774

**Pull request closed without merging**